### PR TITLE
[sysrst_ctrl] Added sysrst_ctrl testbench environment template

### PR DESCRIPTION
**Pull request by patel-madhuri**
Signed-off-by: Madhuri.Patel <madhuri.patel@ensilica.com>

### CODE REVIEW

**Comment by weicaiyang** [comment]
CI is failed due to this error. Please fix it.
> The commit author name 'Madhuri.Patel' contains no space. Use "git config user.name 'Johnny English'" to set your real name, and update the commit with "git rebase -i " and/or "git commit --amend --signoff --reset-author". This command will also sign off your commit indicating agreement to the Contributor License Agreement. See CONTRIBUTING.md for more details.

File: hw/ip/sysrst_ctrl/dv/env/sysrst_ctrl_env.sv [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+class sysrst_ctrl_env extends cip_base_env #(
+    .CFG_T              (sysrst_ctrl_env_cfg),
+    .COV_T              (sysrst_ctrl_env_cov),
+    .VIRTUAL_SEQUENCER_T(sysrst_ctrl_virtual_sequencer),
+    .SCOREBOARD_T       (sysrst_ctrl_scoreboard)
+  );
+  `uvm_component_utils(sysrst_ctrl_env)
+
+  agt1_agent m_agt1_agent;
```
Comment (sriyerg) [inline hw/ip/sysrst_ctrl/dv/env/sysrst_ctrl_env.sv:13]: Please remove this.

File: hw/ip/sysrst_ctrl/dv/tb.sv [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  wire clk, rst_n;
+  wire devmode;
+  wire [NUM_MAX_INTERRUPTS-1:0] interrupts;
+
+  // interfaces
+  clk_rst_if clk_rst_if(.clk(clk), .rst_n(rst_n));
+  pins_if #(NUM_MAX_INTERRUPTS) intr_if(interrupts);
+  pins_if #(1) devmode_if(devmode);
+  tl_if tl_if(.clk(clk), .rst_n(rst_n));
+  agt1_if agt1_if();
```
Comment (sriyerg) [inline hw/ip/sysrst_ctrl/dv/tb.sv:25]: Remove..

File: hw/ip/sysrst_ctrl/dv/sysrst_ctrl_sim_cfg.hjson [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  // RAL spec - used to generate the RAL model.
+  ral_spec: "{proj_root}/hw/ip/sysrst_ctrl/data/sysrst_ctrl.hjson"
+
+  // Import additional common sim cfg files.
+  // TODO: remove imported cfgs that do not apply.
+  import_cfgs: [// Project wide common sim cfg file
+                "{proj_root}/hw/dv/tools/dvsim/common_sim_cfg.hjson",
+                // Common CIP test lists
+                "{proj_root}/hw/dv/tools/dvsim/tests/csr_tests.hjson",
+                "{proj_root}/hw/dv/tools/dvsim/tests/mem_tests.hjson",
```
Comment (sriyerg) [inline hw/ip/sysrst_ctrl/dv/sysrst_ctrl_sim_cfg.hjson:32]: Remove - sysrst ctrl has no memories.

File: hw/ip/sysrst_ctrl/dv/env/sysrst_ctrl_virtual_sequencer.sv [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+class sysrst_ctrl_virtual_sequencer extends cip_base_virtual_sequencer #(
+    .CFG_T(sysrst_ctrl_env_cfg),
+    .COV_T(sysrst_ctrl_env_cov)
+  );
+  `uvm_component_utils(sysrst_ctrl_virtual_sequencer)
+
+  agt1_sequencer agt1_sequencer_h;
```
Comment (sriyerg) [inline hw/ip/sysrst_ctrl/dv/env/sysrst_ctrl_virtual_sequencer.sv:11]: Remove..

File: hw/ip/sysrst_ctrl/dv/env/sysrst_ctrl_env_cfg.sv [hunk 5]
```diff
@@ -0,0 +1,31 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+class sysrst_ctrl_env_cfg extends cip_base_env_cfg #(.RAL_T(sysrst_ctrl_reg_block));
+
+  // ext component cfgs
+  rand agt1_agent_cfg m_agt1_agent_cfg;
```
Comment (sriyerg) [inline hw/ip/sysrst_ctrl/dv/env/sysrst_ctrl_env_cfg.sv:8]: Remove this and the creation in initialize below.

File: hw/ip/sysrst_ctrl/data/sysrst_ctrl_testplan.hjson [hunk 6]
```diff
@@ -0,0 +1,40 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+{
+  name: "sysrst_ctrl"
+  // TODO: remove the common testplans if not applicable
+  import_testplans: ["hw/dv/tools/dvsim/testplans/csr_testplan.hjson",
+                     "hw/dv/tools/dvsim/testplans/mem_testplan.hjson",
```
Comment (sriyerg) [inline hw/ip/sysrst_ctrl/data/sysrst_ctrl_testplan.hjson:8]: This can be removed - sysrst_ctrl has no memories.

File: hw/ip/sysrst_ctrl/dv/tb.sv [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  sysrst_ctrl dut (
+    .clk_i                (clk      ),
+    .rst_ni               (rst_n    ),
+
+    .tl_i                 (tl_if.h2d),
+    .tl_o                 (tl_if.d2h),
+    .alert_rx_i           (alert_rx ),
+    .alert_tx_o           (alert_tx )
+    // TODO: add remaining IOs and hook them
+  );
```
Comment (sriyerg) [inline hw/ip/sysrst_ctrl/dv/tb.sv:39]: Create `sysrst_ctrl_if.sv` under dv/env. Add the rest of the sysrst_ctrl ports to it and connect them in tb. 
Create a virtual interface handle in `sysrst_ctrl_env_cfg` and retrieve handle to the vif via config db.

File: hw/ip/sysrst_ctrl/doc/dv/index.md [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+Please take a look at the link for detailed information on the usage, capabilities, features and known issues.
+Here's how to run a smoke test:
+```console
+$ $REPO_TOP/util/dvsim/dvsim.py $REPO_TOP/hw/ip/sysrst_ctrl/dv/sysrst_ctrl_sim_cfg.hjson -i sysrst_ctrl_smoke
+```
+
+## Testplan
+<!-- TODO: uncomment the line below after adding the testplan.
+Please make sure the testplan is added to `/util/build_docs.py`. -->
+{{</* incGenFromIpDesc "/hw/ip/sysrst_ctrl/data/sysrst_ctrl_testplan.hjson" "testplan" */>}
```
Comment (sriyerg) [inline hw/ip/sysrst_ctrl/doc/dv/index.md:123]: There is a missing `}` at the end.

File: hw/ip/sysrst_ctrl/doc/checklist.md [hunk 9]
```diff
@@ -71,6 +71,14 @@ Security      | [SEC_RND_CNST][]        | N/A         |
 [SEC_CM_DOCUMENTED]:   {{<relref "/doc/project/checklist.md#sec_cm_documented" >}}
 [SEC_RND_CNST]:        {{<relref "/doc/project/checklist.md#sec_rnd_cnst" >}}
 
+### D2S
```
Comment (sriyerg) [inline hw/ip/sysrst_ctrl/doc/checklist.md:74]: Update to this file should not have showed up in this PR.

File: hw/ip/sysrst_ctrl/doc/_index.md [hunk 10]
```diff
@@ -113,9 +113,9 @@ To this end, the hardware samples the raw input values of `pwrb_in`, `key[0,1,2]
 Software can optionally override all output signals, and change the signal polarity of some of the input and output signals.
 The output signal override feature always has higher priority than any of the combo pattern detection mechanisms described above.
 
-The selection of output signals to override, and the override values are programmable and lockable via the {{< regref PIN_LEGAL_CTL >}} register.
-For example, {{< regref PIN_LEGAL_CTL.EC_RST_L_0 >}} to 1 and {{< regref PIN_LEGAL_CTL.EC_RST_L_1 >}} to 0 means that software allows `ec_rst_out_l` to be overridden with logic 0, but not with logic 1.
-If the SW locks the configuration with {{< regref REGWEN >}}, {{< regref PIN_LEGAL_CTL >}} cannot be modified until the next OpenTitan reset.
+The selection of output signals to override, and the override values are programmable and lockable via the {{< regref PIN_ALLOWED_CTL >}} register.
```
Comment (sriyerg) [inline hw/ip/sysrst_ctrl/doc/_index.md:116]: Updates to this file should not have showed up in this PR.

File: hw/ip/sysrst_ctrl/dv/env/sysrst_ctrl_env_pkg.sv [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+  import sysrst_ctrl_ral_pkg::*;
+
+  // macro includes
+  `include "uvm_macros.svh"
+  `include "dv_macros.svh"
+
+  // parameters
+  // TODO: add the names of alerts in order
+  parameter string LIST_OF_ALERTS[] = {};
+  parameter uint   NUM_ALERTS = ;
```
Comment (weicaiyang) [inline hw/ip/sysrst_ctrl/dv/env/sysrst_ctrl_env_pkg.sv:24]: this info can be found in the sysrst_ctrl.hjson. Please update and make this TB pass the compile
```suggestion
  parameter string LIST_OF_ALERTS[] = {"fatal_fault"};
  parameter uint   NUM_ALERTS = 1;
```

**Review by weicaiyang — COMMENTED** [review]
left one comment, and perhaps make this env pass compile, otherwise, LGTM

File: hw/ip/sysrst_ctrl/data/sysrst_ctrl.hjson [hunk 12]
```diff
@@ -322,7 +322,7 @@
         }
         { bits:   "7",
           name:   "flash_wp_l_0",
-          desc:   "1: not allowed; 1: allowed",
```
Comment (sriyerg) [inline hw/ip/sysrst_ctrl/data/sysrst_ctrl.hjson:325]: This change does not belong to the PR. Could you factor this into a separate PR?

File: hw/ip/sysrst_ctrl/data/sysrst_ctrl.hjson [hunk 13]
```diff
@@ -1,873 +0,0 @@
-// Copyright lowRISC contributors.
```
Comment (weicaiyang) [inline hw/ip/sysrst_ctrl/data/sysrst_ctrl.hjson:1]: @patel-madhuri should not delete this file, please just revert the change

**Comment by weicaiyang** [comment]
@patel-madhuri once you address all the comments, please request us to review it again by clicking "re-request review" or ping us in a comment

File: hw/ip/sysrst_ctrl/doc/dv/index.md [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+Please take a look at the link for detailed information on the usage, capabilities, features and known issues.
+Here's how to run a smoke test:
+```console
+$ $REPO_TOP/util/dvsim/dvsim.py $REPO_TOP/hw/ip/sysrst_ctrl/dv/sysrst_ctrl_sim_cfg.hjson -i sysrst_ctrl_smoke
+```
+
+## Testplan
+<!-- TODO: uncomment the line below after adding the testplan.
+Please make sure the testplan is added to `/util/build_docs.py`. -->
+{{</* incGenFromIpDesc "/hw/ip/sysrst_ctrl/data/sysrst_ctrl_testplan.hjson" "testplan" */>}}
```
Comment (weicaiyang) [inline hw/ip/sysrst_ctrl/doc/dv/index.md:123]: This PR has the testplan, no need to comment out the testplan.
```suggestion
{{< incGenFromIpDesc "/hw/ip/sysrst_ctrl/data/sysrst_ctrl_testplan.hjson" "testplan" >}}
```

File: hw/ip/sysrst_ctrl/dv/tb.sv (hunk 7 above)
Comment (weicaiyang) [reply hw/ip/sysrst_ctrl/dv/tb.sv:38]: @patel-madhuri what do you think about this? if you want to address it in a future PR, I think it's fine too.

**Review by weicaiyang — COMMENTED** [review]
Please fix this CI failure and also update the commit message to reflect the changes.
Currently, the commit title is `Reverted sysrst_ctrl.hjson file`, which doesn't seem right.

> Commit has no Signed-off-by line. The commit message must contain a Signed-off-by line that matches the commit author name and email, indicating agreement to the Contributor License Agreement. See CONTRIBUTING.md for more details. You can use "git commit --signoff" to ask git to add this line for you.

File: hw/ip/sysrst_ctrl/dv/tb.sv (hunk 7 above)
Comment (patel-madhuri) [reply hw/ip/sysrst_ctrl/dv/tb.sv:38]: @weicaiyang I will address this change in future  PR

Comment (sriyerg) [reply hw/ip/sysrst_ctrl/dv/tb.sv:38]: Sounds good.

File: hw/ip/sysrst_ctrl/doc/dv/index.md (hunk 14 above)
Comment (patel-madhuri) [reply hw/ip/sysrst_ctrl/doc/dv/index.md]: Uncommented the testplan

File: hw/ip/sysrst_ctrl/dv/env/sysrst_ctrl_env.sv (hunk 1 above)
Comment (patel-madhuri) [reply hw/ip/sysrst_ctrl/dv/env/sysrst_ctrl_env.sv]: As sysrst_ctrl block does not need agent, it is removed from the testbench.

File: hw/ip/sysrst_ctrl/dv/sysrst_ctrl_sim_cfg.hjson (hunk 3 above)
Comment (patel-madhuri) [reply hw/ip/sysrst_ctrl/dv/sysrst_ctrl_sim_cfg.hjson]: Removed mem_tests.hjson

File: hw/ip/sysrst_ctrl/dv/env/sysrst_ctrl_virtual_sequencer.sv (hunk 4 above)
Comment (patel-madhuri) [reply hw/ip/sysrst_ctrl/dv/env/sysrst_ctrl_virtual_sequencer.sv]: Done

File: hw/ip/sysrst_ctrl/dv/env/sysrst_ctrl_env_cfg.sv (hunk 5 above)
Comment (patel-madhuri) [reply hw/ip/sysrst_ctrl/dv/env/sysrst_ctrl_env_cfg.sv]: Done

File: hw/ip/sysrst_ctrl/data/sysrst_ctrl_testplan.hjson (hunk 6 above)
Comment (patel-madhuri) [reply hw/ip/sysrst_ctrl/data/sysrst_ctrl_testplan.hjson]: Removed

File: hw/ip/sysrst_ctrl/doc/dv/index.md (hunk 8 above)
Comment (patel-madhuri) [reply hw/ip/sysrst_ctrl/doc/dv/index.md]: Added  } at the end

File: hw/ip/sysrst_ctrl/doc/_index.md (hunk 10 above)
Comment (patel-madhuri) [reply hw/ip/sysrst_ctrl/doc/_index.md]: Reverted this file

File: hw/ip/sysrst_ctrl/doc/checklist.md (hunk 9 above)
Comment (patel-madhuri) [reply hw/ip/sysrst_ctrl/doc/checklist.md]: Reverted this file

File: hw/ip/sysrst_ctrl/dv/env/sysrst_ctrl_env_pkg.sv (hunk 11 above)
Comment (patel-madhuri) [reply hw/ip/sysrst_ctrl/dv/env/sysrst_ctrl_env_pkg.sv]: Updated parameter value

File: hw/ip/sysrst_ctrl/data/sysrst_ctrl.hjson (hunk 13 above)
Comment (patel-madhuri) [reply hw/ip/sysrst_ctrl/data/sysrst_ctrl.hjson]: Reverted this file

**Review by sriyerg — APPROVED** [review]

File: hw/ip/sysrst_ctrl/doc/dv/index.md [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+We are using our in-house developed [regression tool]({{< relref "hw/dv/tools/README.md" >}}) for building and running our tests and regressions.
+Please take a look at the link for detailed information on the usage, capabilities, features and known issues.
+Here's how to run a smoke test:
+```console
+$ $REPO_TOP/util/dvsim/dvsim.py $REPO_TOP/hw/ip/sysrst_ctrl/dv/sysrst_ctrl_sim_cfg.hjson -i sysrst_ctrl_smoke
+```
+
+## Testplan
+<!-- TODO: uncomment the line below after adding the testplan.
+Please make sure the testplan is added to `/util/build_docs.py`. -->
```
Comment (weicaiyang) [inline hw/ip/sysrst_ctrl/doc/dv/index.md:122]: please remove these 2 lines as well.

Comment (sriyerg) [inline hw/ip/sysrst_ctrl/doc/dv/index.md:122]: The presubmit checks are failing with this message:
```
ERROR 2021/11/18 14:41:03 HTML for testplan has not been generated for /hw/ip/sysrst_ctrl/data/sysrst_ctrl_testplan.hjson (looked at build/docs-generated/hw/ip/sysrst_ctrl/data/sysrst_ctrl_testplan.hjson.testplan)
Total in 2323 ms
Error: Error building site: "/home/vsts/work/1/s/hw/ip/sysrst_ctrl/doc/dv/index.md:123:1": failed to render shortcode "incGenFromIpDesc": failed to process shortcode: "/home/vsts/work/1/s/site/docs/layouts/shortcodes/incGenFromIpDesc.html:30:3": execute of template failed: template: shortcodes/incGenFromIpDesc.html:30:3: executing "shortcodes/incGenFromIpDesc.html" at <readFile $path>: error calling readFile: stat /home/vsts/work/1/s/build/docs-generated/hw/ip/sysrst_ctrl/data/sysrst_ctrl_testplan.hjson.testplan: no such file or directory
Error building site
```

The reason is this comment on line 122 needs to be satisfied. Please update `util/build_docs.py` and add entry in there so that it can find the sysrst CTRL testplan correctly.

**Review by weicaiyang — APPROVED** [review]
LGTM, thanks

**Comment by sriyerg** [comment]
The failiing FPGA CI check does not seem related to this PR. I will merge it.
